File: CSharp/SearchingAlgorithms/BinarySearch.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace CSharp
{
    public static class BinarySearch
    {
        public static int Search<T>(IEnumerable<T> arr, T target) where T: IComparable
        {
            var left = 0;
            var right = arr.Count() - 1;

            int mid;

            while (left <= right)
            {
                mid = (left + right) / 2;
                var current = arr.ElementAt(mid);

                if (current.Equals(target))
                {
                    return mid;
                }
                else if (current.CompareTo(target) < 0)
                {
                    left = mid + 1;
                }
                else
                {
                    right = mid - 1;
                }
            }

            return -1;

        }
    
        // check how many times the smaller string m appears inside the bigger one s
        public static int NaiveStringSearch(string s, string m)
        {
            var count = 0;
            for (var i = 0; i < s.Length; i++)
            {
                if (s[i] == m[0])
                {
                    for (var j = 0; j < m.Length; j++)
                    {
                        if (s[i + j] != m[j])
                            break;
                    }
                    count++;
                }
            }
            return count;
        }
    }
}

File: JS/LeetC/g75/binarySearch.js
function search(nums, target) {

  let start = 0;
  let end = nums.length - 1;

  while (start < end) {
    let mid = Math.floor(start + (end - start) / 2);

    let current = nums[mid];

    if (target < current) {
      end = mid - 1;
    } else if (target > current) {
      start = mid + 1;
    } else {
      return mid;
    }
  }

  return -1;
}

console.log(search([-1, 0, 3, 5, 9, 12], 9));
